File: schnorrkel-bench/schnorrkel.rs
use criterion::{criterion_group, criterion_main, Criterion};
use rand::{distributions::Alphanumeric, thread_rng, Rng};
use schnorrkel::{
    context::SigningContext, signing_context, ExpansionMode, MiniSecretKey, PublicKey, SecretKey,
    Signature, PUBLIC_KEY_LENGTH, SECRET_KEY_LENGTH, SIGNATURE_LENGTH,
};
use sha2::{Digest, Sha256};

struct Input {
    context: SigningContext,
    context_bytes: [u8; 32],

    seckey: SecretKey,
    seckey_bytes: [u8; SECRET_KEY_LENGTH],

    pubkey: PublicKey,
    pubkey_bytes: [u8; PUBLIC_KEY_LENGTH],

    messages: Vec<[u8; 32]>,

    signatures: Vec<Signature>,
    signatures_bytes: Vec<[u8; SIGNATURE_LENGTH]>,
}

fn make_hash(bytes: &[u8]) -> [u8; 32] {
    let mut hasher = Sha256::new();
    hasher.update(bytes);
    hasher.finalize().into()
}

fn generate_input() -> Vec<Input> {
    let seed = match std::env::var("SEED") {
        Ok(seed) => seed,
        Err(std::env::VarError::NotPresent) => {
            let mut rng = thread_rng();
            std::iter::repeat(())
                .take(32)
                .map(|()| rng.sample(Alphanumeric))
                .map(char::from)
                .collect()
        }
        Err(error) => panic!("Failed to get seed: {}", error),
    };
    println!("Input seed: {}", seed);
    let seed = make_hash(seed.as_bytes());

    std::iter::repeat_with(|| {
        let seed = make_hash(&seed);
        let context_bytes = seed.clone();
        let context = signing_context(&context_bytes);

        let seed = make_hash(&seed);
        let mini_seckey = MiniSecretKey::from_bytes(&seed).unwrap();
        let keypair = mini_seckey.expand_to_keypair(ExpansionMode::Uniform);

        let messages = std::iter::repeat_with(|| {
            let seed = make_hash(&seed);
            seed.clone()
        })
        .take(10)
        .collect::<Vec<[u8; 32]>>();

        let signatures = messages
            .iter()
            .map(|message| keypair.secret.sign(context.bytes(message), &keypair.public))
            .collect::<Vec<Signature>>();
        let signatures_bytes = signatures
            .iter()
            .map(|signature| signature.to_bytes())
            .collect();

        Input {
            context,
            context_bytes,

            seckey: keypair.secret.clone(),
            seckey_bytes: keypair.secret.to_bytes(),

            pubkey: keypair.public.clone(),
            pubkey_bytes: keypair.public.to_bytes(),

            messages,

            signatures,
            signatures_bytes,
        }
    })
    .take(100)
    .collect()
}

fn schnorrkel_benchmark(c: &mut Criterion) {
    let input = generate_input();

    let mut group = c.benchmark_group("steateless");
    group.bench_with_input("sign", &input, |b, input| {
        b.iter(|| {
            for item in input {
                let context = signing_context(&item.context_bytes);
                let seckey = SecretKey::from_bytes(&item.seckey_bytes).unwrap();
                let pubkey = PublicKey::from_bytes(&item.pubkey_bytes).unwrap();
                let _signature = seckey.sign(context.bytes(&item.messages[0]), &pubkey);
            }
        })
    });
    group.bench_with_input("verify", &input, |b, input| {
        b.iter(|| {
            for item in input {
                let context = signing_context(&item.context_bytes);
                let pubkey = PublicKey::from_bytes(&item.pubkey_bytes).unwrap();
                let signature = Signature::from_bytes(&item.signatures_bytes[0]).unwrap();
                pubkey
                    .verify(context.bytes(&item.messages[0]), &signature)
                    .unwrap();
            }
        })
    });
    group.finish();

    let mut group = c.benchmark_group("stateful");
    group.bench_with_input("sign", &input, |b, input| {
        b.iter(|| {
            for item in input.iter().take(10) {
                for message in item.messages.iter() {
                    let _signature = item.seckey.sign(item.context.bytes(message), &item.pubkey);
                }
            }
        })
    });
    group.bench_with_input("verify", &input, |b, input| {
        b.iter(|| {
            for item in input.iter().take(10) {
                for (message, signature) in item.messages.iter().zip(item.signatures.iter()) {
                    item.pubkey
                        .verify(item.context.bytes(message), signature)
                        .unwrap()
                }
            }
        })
    });
    group.finish();
}

criterion_group!(benches, schnorrkel_benchmark);
criterion_main!(benches);

File: schnorrkel-bench/schnorrkel.js
const assert = require("assert");
const crypto = require("crypto");
const { Suite } = require("benchmark");
const napi = require("../schnorrkel-napi");
const wasm = require("../schnorrkel-wasm/pkg");

function make_hash(input) {
  return crypto.createHash("sha256").update(input).digest();
}

function generate_seed() {
  let seed = "";
  while (seed.length < 32) {
    const byte = crypto.randomBytes(1)[0];
    if (
      ("0" >= byte && "9" <= byte) ||
      ("a" >= byte && "z" <= byte) ||
      ("A" >= byte && "Z" <= byte)
    ) {
      seed += String.fromCharCode(byte);
    }
  }
  return seed;
}

function generate_input() {
  let seed = process.env.SEED || generate_seed();
  console.log(`Input seed: ${seed}`);
  seed = make_hash(seed);

  return new Array(100).fill(null).map(() => {
    seed = make_hash(seed);
    const context_bytes = Buffer.from(seed);
    const context = napi.stateful.create_context(context_bytes);

    seed = make_hash(seed);
    const obj = napi.stateful.generate_pair(seed);
    const { seckey, seckey_bytes, pubkey, pubkey_bytes } = obj;

    const messages = new Array(10).fill(null).map(() => {
      seed = make_hash(seed);
      return Buffer.from(seed);
    });

    const signatures = [];
    const signatures_bytes = [];
    for (const message of messages) {
      const obj = napi.stateful.sign(context, seckey, pubkey, message);
      const { signature, signature_bytes } = obj;
      signatures.push(signature);
      signatures_bytes.push(signature_bytes);
    }

    return {
      context,
      context_bytes,

      seckey,
      seckey_bytes,

      pubkey,
      pubkey_bytes,

      messages,

      signatures,
      signatures_bytes,
    };
  });
}

const input = generate_input();
new Suite()
  .add("napi/stateful/sign", () => {
    for (let i = 0; i < 10; ++i) {
      const item = input[i];
      for (const message of item.messages) {
        napi.stateful.sign(item.context, item.seckey, item.pubkey, message);
      }
    }
  })
  .add("napi/stateful/verify", () => {
    for (let i = 0; i < 10; ++i) {
      const item = input[i];
      for (let j = 0; j < item.messages.length; ++j) {
        const message = item.messages[j];
        const signature = item.signatures[j];
        assert(
          napi.stateful.verify(item.context, item.pubkey, message, signature)
        );
      }
    }
  })
  .add("napi/stateless/sign", () => {
    for (const item of input) {
      napi.stateless.sign(
        item.context_bytes,
        item.seckey_bytes,
        item.pubkey_bytes,
        item.messages[0]
      );
    }
  })
  .add("napi/stateless/verify", () => {
    for (const item of input) {
      assert(
        napi.stateless.verify(
          item.context_bytes,
          item.pubkey_bytes,
          item.messages[0],
          item.signatures_bytes[0]
        )
      );
    }
  })
  .add("wasm/stateless/sign", () => {
    for (const item of input) {
      wasm.stateless_sign(
        item.context_bytes,
        item.seckey_bytes,
        item.pubkey_bytes,
        item.messages[0]
      );
    }
  })
  .add("wasm/stateless/verify", () => {
    for (const item of input) {
      assert(
        wasm.stateless_verify(
          item.context_bytes,
          item.pubkey_bytes,
          item.messages[0],
          item.signatures_bytes[0]
        )
      );
    }
  })
  .on("cycle", (event) => console.log(String(event.target)))
  .run();
